### PR TITLE
runtime: fix the bug of func countFiles

### PR DESCRIPTION
**Pull request by gaohuatao-1**
When the total number of files observed is greater than limit, return (-1, err). When the returned err is not nil, the func countFiles should return -1.

Fixes:#9780

### CODE REVIEW

File: src/runtime/virtcontainers/mount.go [hunk 1]
```diff
@@ -371,7 +371,7 @@ func countFiles(path string, limit int) (numFiles int, err error) {
 	// First, Check to see if the path exists
 	file, err := os.Stat(path)
 	if os.IsNotExist(err) {
-		return 0, err
+		return -1, err
```
[the remark this answers is not in the document]
Comment (gaohuatao-1) [reply src/runtime/virtcontainers/mount.go:374]: > Why we need this? IMHO, we test if the err is empty firstly. If no, a process of handling error is required. Thus, the first return value should not be used.

To be consistent with the code comments: "break and return -1". On the other hand, in order to distinguish from the two situations where no file under path  and encounting an error.

[the remark this answers is not in the document]
Comment (amshinde) [reply src/runtime/virtcontainers/mount.go:374]: @gaohuatao-1  Agree with @justxuewei, Also according to comments -1 is returned only in case of linmit being exceeded.

**Comment by justxuewei** [comment]
Hi @gaohuatao-1, I think this PR couldn't get merged. For this method, there are three cases:

1. Success: return `numFiles, nil`;
2. Failure
    1. the total number of files observed is greater than limit: return `-1, nil`;
    2. other errors: return `0, err`: if `err != nil`, the first value should be ignored.

**Comment by gaohuatao-1** [comment]
> Hi @gaohuatao-1, I think this PR couldn't get merged. For this method, there are three cases:
> 
> 1. Success: return `numFiles, nil`;
> 2. Failure
>    
>    1. the total number of files observed is greater than limit: return `-1, nil`;
>    2. other errors: return `0, err`: if `err != nil`, the first value should be ignored. It is exactly the same, wether it is -1 or 0.
> 
> > in order to distinguish from the two situations where no file under path and encounting an error.
> 
> What is "encounting an error"? I don't think that changing `0` to `-1` could achieve your goal of distinguishing errors.
> 
> WDYT? Or could you elaborate a bit about why you need these changes?



> elaborate

Thanks for your reply. In the issue, I descibed the problem that the func can not achieve the goal you mentioned: the total number of files observed is greater than limit: return `-1, nil`;
Using this func to do a test:
image
The test code is:
```
package main

import (
	"fmt"
	"os"
	"path/filepath"
)

func main() {
	path := os.Args[1]
	num, err := countFiles(path, 8)
	fmt.Println(num, err)
}

// countFiles will return the number of files within a given path. If the total number of
// files observed is greater than limit, break and return -1
func countFiles(path string, limit int) (numFiles int, err error) {

	// First, Check to see if the path exists
	file, err := os.Stat(path)
	if os.IsNotExist(err) {
		return 0, err
	}

	// Special case if this is just a file, not a directory:
	if !file.IsDir() {
		return 1, nil
	}

	files, err := os.ReadDir(path)
	if err != nil {
		return 0, err
	}

	for _, file := range files {
		if file.IsDir() {
			inc, err := countFiles(filepath.Join(path, file.Name()), (limit - numFiles))
			if err != nil {
				return numFiles, err
			}
			numFiles = numFiles + inc
		} else {
			numFiles++
		}
		if numFiles > limit {
			return -1, nil
		}
	}
	return numFiles, nil
}
```

**Comment by justxuewei** [comment]
I see. Shall we change like the following?

```go
func countFiles(path string, limit int) (numFiles int, err error) {

	// First, Check to see if the path exists
	file, err := os.Stat(path)
	if os.IsNotExist(err) {
		return 0, err
	}

	// Special case if this is just a file, not a directory:
	if !file.IsDir() {
		return 1, nil
	}

	files, err := os.ReadDir(path)
	if err != nil {
		return 0, err
	}

	for _, file := range files {
		if file.IsDir() {
			inc, err := countFiles(filepath.Join(path, file.Name()), (limit - numFiles))
			if err != nil {
				return numFiles, err
			}
+			// exceeded limit
+			if numFiles == -1 {
+				return -1, nil
+			}
			numFiles = numFiles + inc
		} else {
			numFiles++
		}
		if numFiles > limit {
			return -1, nil
		}
	}
	return numFiles, nil
}
```

**Comment by gaohuatao-1** [comment]
> I see. Shall we change like the following? Could you help me try this snippet?
> 
> ```go
> func countFiles(path string, limit int) (numFiles int, err error) {
> 
> 	// First, Check to see if the path exists
> 	file, err := os.Stat(path)
> 	if os.IsNotExist(err) {
> 		return 0, err
> 	}
> 
> 	// Special case if this is just a file, not a directory:
> 	if !file.IsDir() {
> 		return 1, nil
> 	}
> 
> 	files, err := os.ReadDir(path)
> 	if err != nil {
> 		return 0, err
> 	}
> 
> 	for _, file := range files {
> 		if file.IsDir() {
> 			inc, err := countFiles(filepath.Join(path, file.Name()), (limit - numFiles))
> 			if err != nil {
> 				return numFiles, err
> 			}
> +			// exceeded limit
> +			if inc == -1 {
> +				return -1, nil
> +			}
> 			numFiles = numFiles + inc
> 		} else {
> 			numFiles++
> 		}
> 		if numFiles > limit {
> 			return -1, nil
> 		}
> 	}
> 	return numFiles, nil
> }
> ```

The returned value is `5, nil`, it is still wrong.

**Comment by gaohuatao-1** [comment]
> I see. Shall we change like the following? Could you help me try this snippet?
> 
> ```go
> func countFiles(path string, limit int) (numFiles int, err error) {
> 
> 	// First, Check to see if the path exists
> 	file, err := os.Stat(path)
> 	if os.IsNotExist(err) {
> 		return 0, err
> 	}
> 
> 	// Special case if this is just a file, not a directory:
> 	if !file.IsDir() {
> 		return 1, nil
> 	}
> 
> 	files, err := os.ReadDir(path)
> 	if err != nil {
> 		return 0, err
> 	}
> 
> 	for _, file := range files {
> 		if file.IsDir() {
> 			inc, err := countFiles(filepath.Join(path, file.Name()), (limit - numFiles))
> 			if err != nil {
> 				return numFiles, err
> 			}
> +			// exceeded limit
> +			if inc == -1 {
> +				return -1, nil
> +			}
> 			numFiles = numFiles + inc
> 		} else {
> 			numFiles++
> 		}
> 		if numFiles > limit {
> 			return -1, nil
> 		}
> 	}
> 	return numFiles, nil
> }
> ```

Yes, this modification returns -1, nil.

**Comment by justxuewei** [comment]
> Yes, this modification returns -1, nil.

I think you could modify it to this ;)

**Comment by gaohuatao-1** [comment]
> > Yes, this modification returns -1, nil.
> 
> I think you could modify it to this ;) /cc @amshinde

PTAL! The function caller needs to judge returned err, I think.

**Review by justxuewei — APPROVED** [review]
> PTAL! The function should return 0 when err is not nil, I think.

Yeah, good catch! And lgtm, thanks!

**Review by amshinde — APPROVED** [review]
